src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/AdaptedPipeline.cs

```diff
@@ -3,37 +3,41 @@
 
 using System;
 using System.IO;
+using System.IO.Pipelines;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
 using Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure;
+using MemoryPool = Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure.MemoryPool;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal
 {
     public class AdaptedPipeline : IDisposable
     {
+        private const int AllocBufferSize = 2048;
+
         private readonly Stream _filteredStream;
 
         public AdaptedPipeline(
             string connectionId,
             Stream filteredStream,
+            Pipe pipelineFactory,
             MemoryPool memory,
-            IKestrelTrace logger,
-            IThreadPool threadPool,
-            IBufferSizeControl bufferSizeControl)
+            IKestrelTrace logger)
         {
-            SocketInput = new SocketInput(memory, threadPool, bufferSizeControl);
-            SocketOutput = new StreamSocketOutput(connectionId, filteredStream, memory, logger);
+            Input = pipelineFactory;
+            Output = new StreamSocketOutput(connectionId, filteredStream, memory, logger);
 
             _filteredStream = filteredStream;
         }
 
-        public SocketInput SocketInput { get; }
+        public Pipe Input { get; }
 
-        public ISocketOutput SocketOutput { get; }
+        public ISocketOutput Output { get; }
 
         public void Dispose()
         {
-            SocketInput.Dispose();
+            Input.CompleteReader();
+            Input.CompleteWriter();
```

Comment:
Don't we wait for ReadInputAsync() to complete before calling this Dispose() method? If so, isn't guaranteed that CompleteWriter() was already called?